Repository GCS plugin new client library

albertzaharovits opened the pull request:
This does away with the `com.google.api-client:google-api-client:1.23` and replaces it with `com.google.cloud:google-cloud-storage:1.27.0` . It also changes security permissions for this plugin.

The API of the new library is indeed easier to use, but the configuration requires uncovering internals which most likely will change in the new versions (because the internals are actually the old library that will go away). This should be easy to change when it will happen, although some configuration options might go away (eg `applicationName` or `projectId`)

DISCLAIMER: This is not turnkey ready. The REST integ test `./gradlew :plugins:repository-gcs:integTest` does not work so I have tested the plugin manually with my Google Cloud account. I have tested it using the Service Account JSON file.

The reason the integration test does not work is that the Service Account JSON file, that we contrive to direct the cluster to the mock cloud, does not allow to override the `oauth/token` endpoint, which is the first endpoint called in the flow. So, even if the subsequent API calls should be compatible with our existing mock, to obtain the token, there is a call to 'googleapis' which obviously gets 'unauthorized'. I don't think this is intended, just a sloppiness in the transition.

Another worry is that the [Application Default Credentials](https://cloud.google.com/docs/authentication/production) used to authenticate when ES runs on a VM provisioned inside google cloud, reads files from various places (ENV vars, user home, ... other heuristics), so a snap-in `security-permission.policy` is not an option. We could mimic what we do with our config dir, but in this case we would have to duplicate their credentials search logic.

Closes: https://github.com/elastic/elasticsearch/issues/29259

elasticmachine commented:
Pinging @elastic/es-distributed

albertzaharovits commented on plugins/repository-gcs/build.gradle:
```diff
@@ -22,40 +17,59 @@ import java.security.KeyPairGenerator
-  compile "commons-codec:commons-codec:${versions.commonscodec}"
-  compile "com.google.http-client:google-http-client:${versions.google}"
-  compile "com.google.http-client:google-http-client-jackson2:${versions.google}"
+  compile 'com.google.cloud:google-cloud-storage:1.26.0'
```
👍

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -19,74 +19,92 @@
+                .setMaxRetryDelay(Duration.ofMillis(6000))
+                .setTotalTimeout(Duration.ofMillis(900000))
+                .setRetryDelayMultiplier(1.5d)
+                .setJittered(true)
```
👍  removed

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -19,74 +19,92 @@
      * @return a Client instance that can be used to manage Storage objects
      */
-    public Storage createClient(final String clientName) throws Exception {
+    public Storage createClient(String clientName) throws GeneralSecurityException, IOException {
```
Yes, sorry for this. The line changed because the exception list changed, and the IDE automatically removed the final from parameters (as this is not a rule in my IDE config).

albertzaharovits commented on plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java:
```diff
@@ -19,289 +19,565 @@
+            optionMethod.setAccessible(true);
+            assert StorageRpc.Option.PREFIX.equals(optionMethod.invoke(prefixFilter)) : "Only the prefix filter is mocked";
+            final Method valueMethod = BlobListOption.class.getSuperclass().getDeclaredMethod("getValue");
+            valueMethod.setAccessible(true);
```
These, unfortunately require the `suppressAccessChecks` permission. (Answer to review comment: https://github.com/elastic/elasticsearch/pull/30168/files#r184983763)

tlrx replied:
@albertzaharovits and I talked about this, and we figured out that this is not the only place where the permission is needed but the main code requires it too.

I still think that we should not add a specific permission for tests in this case because it can hide other places where a security exception could be thrown where in fact we should work to remove this permission in the main plugin. I also think that in this case Mockito makes things harder to maintain and to read where we can instead [use POJOs to mock the Google Storage](https://github.com/tlrx/elasticsearch/blob/1f40ea509406b14ac4300828b27b9418562eee59/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/MockStorage.java) but with this option we have to pay the price of adding two utility classed in foreign packages like `com.google.storage`. It's like to choose between two bad options but this is an interesting point and I'd like to know @elastic/es-core-infra opinion

rjernst replied:
Anything is better than allowing/using setAccessible. Creating utility classes in a foreign package is not great, but definitely preferable, IMO.

tlrx replied:
Thanks @rjernst and core/infra team.

albertzaharovits commented on plugins/repository-gcs/src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -20,10 +20,15 @@
   permission java.lang.RuntimePermission "accessDeclaredMembers";
   permission java.lang.RuntimePermission "setFactory";
+  permission java.lang.RuntimePermission "accessClassInPackage.sun.misc";
   permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
```
Yes, unfortunately suppressAccessChecks is required in the `MockStorage`. I have left a comment pointing to the two culprit lines. There might be required inside the library too, I haven't checked.

tlrx replied:
> Yes, unfortunately suppressAccessChecks is required in the MockStorage.

I think you can add specific permissions like this in a src/test/resources/plugin-security.policy file. Did you try that?

albertzaharovits commented on plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageServiceTests.java:
```diff
@@ -19,79 +19,64 @@
-import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 public class GoogleCloudStorageServiceTests extends ESTestCase {
```
👍

albertzaharovits commented on plugins/repository-gcs/src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -20,10 +20,15 @@
+
+  //add read file permissions for the gcloud credentials
+  //eg. new File(System.getProperty("user.home"), ".config/gcloud")
+  //permission java.io.FilePermission "<<ALL FILES>>", "read";
```
> I think that using the default credentials is just broken

I agree, but I think this should be part of another docs? PR. This PR is already quite beefy.

Alternatively, if we aim for the maximum of convenience we could grant `<<ALL FILES>>` permission to the plugin, **but** all the `doPrivileged` calls have to narrow the privileges required. I have left a comment: https://github.com/elastic/elasticsearch/issues/23992#issuecomment-385930499

For now, I have kept the original behavior (which is tripping the permission exception while reading credentials from the FS), but logged:
```
logger.warn("\"Application Default Credentials\" are not supported out of the box."
+                    + " Additional file system permissions have to be granted to the plugin.");
```

tlrx replied:
> I agree, but I think this should be part of another docs? PR. This PR is already quite beefy.

Yes, I didn't expect this PR to be so large to be honest. Let's address this in another PR.

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -204,13 +212,22 @@ InputStream readBlob(String blobName) throws IOException {
+                    }
+                }
+            }
+            assert blobSize == bytesWritten : "InputStream unexpected size, expected [" + blobSize + "] got [" + bytesWritten + "]";
```
Thanks again for the detailed suggestion. I have used it as it is.

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -19,74 +19,90 @@
+            final ServiceAccountCredentials serviceAccountCredentials = clientSettings.getCredential();
+            if (Strings.hasLength(clientSettings.getTokenUri().toString())) {
+                storageOptionsBuilder
+                        .setCredentials(serviceAccountCredentials.toBuilder().setTokenServerUri(clientSettings.getTokenUri()).build());
```
override token uri, using client setting from `elasticsearch.yml`

tlrx replied:
Maybe split this in multiple lines for a better readability?
```
            final URI tokenUri = clientSettings.getTokenUri();
            if (tokenUri != null) {
                // Rebuild the service account credentials in order to use a custom Token url.
                // This is mostly used for testing purpose.
                ServiceAccountCredentials.Builder credentialsBuilder = serviceAccountCredentials.toBuilder();
                credentialsBuilder.setTokenServerUri(tokenUri);
                storageOptionsBuilder.setCredentials(credentialsBuilder.build());
            } else {
                storageOptionsBuilder.setCredentials(serviceAccountCredentials);
            }
```

albertzaharovits replied:
hope this fills the bill:
```
            ServiceAccountCredentials serviceAccountCredentials = clientSettings.getCredential();
            // override token server URI
            final URI tokenServerUri = clientSettings.getTokenUri();
            if (Strings.hasLength(tokenServerUri.toString())) {
                // Rebuild the service account credentials in order to use a custom Token url.
                // This is mostly used for testing purpose.
                serviceAccountCredentials = serviceAccountCredentials.toBuilder().setTokenServerUri(tokenServerUri).build();
            }
            storageOptionsBuilder.setCredentials(serviceAccountCredentials);
```

tlrx replied:
Thanks!

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -204,14 +187,35 @@ InputStream readBlob(String blobName) throws IOException {
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream(Math.toIntExact(blobSize));
+            Streams.copy(inputStream, baos);
+            SocketAccess.doPrivilegedVoidIOException(() -> storage.create(blobInfo, baos.toByteArray()));
+        }
```
This is a legit optimization according to the docs, but actually it is necessary to get around the library, which [does not use the port](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/52b727aef88ae76984aa3c02b4d7067e198d34b7/google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java#L720) for the new connections it opens for the "resumable" upload.
This means the REST Integ tests will try to upload files using the default port 80, but the fixture cannot bind this port.
The multipart upload builds the URL correctly.

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -241,163 +246,57 @@ void deleteBlobsByPrefix(String prefix) throws IOException {
+        if (blobNames.size() < 5) {
+            for (final String blobName : blobNames) {
+                deleteBlob(blobName);
+            }
```
This "optimization" is not justifiable per se, but it is required to get around another URL building blunder inside the google cloud library. [Batched requests don't honor the hostname + port configuration](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/52b727aef88ae76984aa3c02b4d7067e198d34b7/google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java#L191). All batched request paths are hardcoded to `www.googleapis.com` 😢 
REST integ tests don't delete more than 3 files in a single batch, so this code branch is required to make the integ tests work.

albertzaharovits commented on plugins/repository-gcs/qa/google-cloud-storage/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageTestServer.java:
```diff
@@ -224,33 +229,150 @@ public Response handle(final String method,
+                    }
+                }
+                // Read line by line ?both? parts of the multipart
+                try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStreamBody, StandardCharsets.ISO_8859_1))) {
```
Mixing UTF-8 encoded text (variable length encoding) with binary data in the same entity is a sure source of bugs and headaches 💯 
Decoding text as ISO_8859_1 simplifies the parsing.

albertzaharovits commented:
There has been some pain to get the REST integ tests (as slim as they are) to behave. Fortunately, 
`gradle :plugins:repository-gcs:check` passes with the local fixture as well as real GCS credentials. I have also addressed all the comments. I have not changed the `Application Default Credentials` behavior (https://github.com/elastic/elasticsearch/issues/23992).

@tlrx Please, may you take another look?

tlrx commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -241,163 +246,57 @@ void deleteBlobsByPrefix(String prefix) throws IOException {
      */
     void deleteBlobs(Collection<String> blobNames) throws IOException {
-        if (blobNames == null || blobNames.isEmpty()) {
+        if ((blobNames == null) || blobNames.isEmpty()) {
```
Note: There's no strict rule but we don't always add parenthesis in `if` conditions. In this case for example, it does not bring much.

tlrx commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -241,163 +246,57 @@ void deleteBlobsByPrefix(String prefix) throws IOException {
-            client.objects().delete(bucket, sourceBlob).execute();
-            return null;
+            final CopyWriter copyWriter = storage.copy(request);
+            copyWriter.getResult();
```
I'd change this for: `storage.copy(request).getResult();`

tlrx commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -241,163 +246,57 @@ void deleteBlobsByPrefix(String prefix) throws IOException {
+            copyWriter.getResult();
+            final boolean deleted = storage.delete(sourceBlobId);
+            if (deleted == false) {
+                throw new IOException("Failed to move source [" + sourceBlobName + "] to target [" + targetBlobName + "].");
```
Nit: Exception messages and log traces usually have no dot at the end, except if they contain more than 1 sentence.

albertzaharovits replied:
I haven't noticed this, thanks.

tlrx commented on plugins/repository-gcs/src/main/plugin-metadata/plugin-security.policy:
```diff
@@ -20,9 +20,10 @@
 grant {
   permission java.lang.RuntimePermission "accessDeclaredMembers";
   permission java.lang.RuntimePermission "setFactory";
+  permission java.lang.RuntimePermission "accessClassInPackage.sun.misc";
```
Can you add a comment why it needs to be added?

albertzaharovits replied:
I have removed this permission altogether looks like it's not required (did some manual testing to convince my eyes).

One would say populating the permission list is easy peasy: when you have an access exception you append the list. I don't recall the specific failure and now the failure went away. I would guess it went away when updating some dependcy libs...

tlrx commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettings.java:
```diff
@@ -44,11 +48,35 @@
+     * An override for the Storage endpoint to connect to. Deprecated, use host
+     * setting.
+     */
     static final Setting.AffixSetting<String> ENDPOINT_SETTING = Setting.affixKeySetting(PREFIX, "endpoint",
```
Why does it need to be deprecated?

albertzaharovits replied:
The only reason is for the setting name to follow the rename of the sibling setting name in the library.
You can indeed set the protocol and port using the `host` setting, but, there are some cases when new connections don't follow the port specification. My impression is that the client library in fact expects this to only specify the hostname, yet it is not enforced or tested yet, and I think we should relay this "suggestion". No strong bias, though.

albertzaharovits replied:
I did, `endpoint` is the sole hostname setting .

tlrx commented:
I think we can get rid of the specific permission by using a different implementation of MockStorage that does not use mockito and does not require the additionnal permission.

I implemented it here: https://github.com/tlrx/elasticsearch/commit/1f40ea509406b14ac4300828b27b9418562eee59, I think this is how MockStorage should be.

albertzaharovits commented:
Thank you Tanguy for the prompt and diligent review! I have addressed it all.

These rogue permissions have been removed:
```
permission java.lang.RuntimePermission "accessClassInPackage.sun.misc";
permission java.net.URLPermission "https://www.googleapis.com/-", "*:*";
permission java.net.URLPermission "https://accounts.google.com/-", "*:*";
```
 Apologies, I don't recall what the errors were when they came to be.

I haven't however adopted the `MockStorage` you propose:
> I think we can get rid of the specific permission by using a different implementation of MockStorage > that does not use mockito and does not require the additionnal permission.
> I implemented it here: tlrx/elasticsearch@1f40ea5, I think this is how MockStorage should be.

Just to check I understand, generally speaking we should avoid writing classes inside foreign packages and use reflection to gain access to hidden members. However, this special case (`com.google.cloud.storage.StorageRpcOptionUtils`) was mandated by the goal to get rid of the `suppressAccessChecks` permission, assuming the permission was required **only** by the test code. Is this correct? If the permission is required by the `src/main` code, and therefore cannot be removed, are classes written outside the `org.elasticsearch` considered a "code smell"?

tlrx commented on plugins/repository-gcs/qa/google-cloud-storage/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageTestServer.java:
```diff
@@ -224,33 +229,152 @@ public Response handle(final String method,
         // https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload
         handlers.insert("PUT " + endpoint + "/upload/storage/v1/b/{bucket}/o", (params, headers, body) -> {
-            String objectId = params.get("upload_id");
+            final String objectId = params.get("upload_id");
```
Do we still need this handler?

albertzaharovits replied:
Yes it is still required when resumable upload is used.

albertzaharovits commented on plugins/repository-gcs/qa/google-cloud-storage/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageTestServer.java:
```diff
@@ -224,33 +229,152 @@ public Response handle(final String method,
+                 * KEwE3bU4TuyetBgQIghmUw
+                 * --__END_OF_PART__--
+                 */
+                String boundary = "__END_OF_PART__";
```
I have pondered it too, but I think they should be kept separate. The "theoretical" reason is that for the `multipart/related` content type the entities are tied together. The first entity contains metadata about the data from the second entity. There is an implied order, and the number of entities is fixed. For the `multipart/mixed` all entities are homogeneous.
The "practical" reason is that parsing these is intricate and grouping them together will make them more so.

The only common denominator I see is parsing the headers of each part.

albertzaharovits commented on plugins/repository-gcs/qa/google-cloud-storage/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageTestServer.java:
```diff
@@ -301,6 +425,37 @@ public Response handle(final String method,
+        // Rewrite Object
+        //
+        // https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite
+        handlers.insert("POST " + endpoint + "/storage/v1/b/{srcBucket}/o/{src}/rewriteTo/b/{destBucket}/o/{dest}",
```
I have done so although, it looks like `copyTo` has fallen into disgrace in favor of `rewriteTo` but I cannot guarantee it's not used anymore, or that it will not be used again in a minor version update of the lib.

albertzaharovits commented on plugins/repository-gcs/qa/google-cloud-storage/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageTestServer.java:
```diff
@@ -398,42 +553,42 @@ public Response handle(final String method,
                         line = reader.readLine();
                         byte[] batchedBody = new byte[0];
-                        if (line != null || line.startsWith("--" + boundary) == false) {
+                        if ((line != null) || (line.startsWith("--" + boundary) == false)) {
```
Apologies, I will reign in my IDE!

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -244,160 +278,53 @@ void deleteBlobs(Collection<String> blobNames) throws IOException {
-
-        if (blobNames.size() == 1) {
-            deleteBlob(blobNames.iterator().next());
+        if (blobNames.size() < 3) {
```
Agreed, filed:
https://github.com/GoogleCloudPlatform/google-cloud-java/issues/3254
https://github.com/GoogleCloudPlatform/google-cloud-java/issues/3255

I like the work around you have proposed as it does not have an impact on the production code, but it's still messy and deals with internals. Hope we can remove this soon.

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettings.java:
```diff
@@ -66,13 +94,23 @@
     /** Name used by the client when it uses the Google Cloud JSON API. **/
     static final Setting.AffixSetting<String> APPLICATION_NAME_SETTING = Setting.affixKeySetting(PREFIX, "application_name",
-        key -> new Setting<>(key, "repository-gcs", s -> s, Setting.Property.NodeScope));
+            key -> new Setting<>(key, "elasticsearch-repository-gcs", Function.identity(), Setting.Property.NodeScope,
```
Ok. Originally, I had taken the liberty to change this because it is prefixed internally by the library and appends the version too, and I though it's not that important, although who I'm to say it might not get untangled on the server side.

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java:
```diff
@@ -19,74 +19,94 @@
-        Storage.Builder storage = new Storage.Builder(transport, JacksonFactory.getDefaultInstance(), requestInitializer);
-        if (Strings.hasLength(clientSettings.getApplicationName())) {
-            storage.setApplicationName(clientSettings.getApplicationName());
+        final NetHttpTransport netHttpTransport = GoogleNetHttpTransport.newTrustedTransport();
```
I have coded smth similar. I have not copied it verbatim because `URL#getFile` is a forbiddenApi and a URI conversion was required.
```
    private static HttpTransport createHttpTransport(final String endpoint) throws Exception {
        final NetHttpTransport.Builder builder = new NetHttpTransport.Builder();
        // requires java.lang.RuntimePermission "setFactory"
        builder.trustCertificates(GoogleUtils.getCertificateTrustStore());
        if (Strings.hasLength(endpoint)) {
            final URL endpointUrl = URI.create(endpoint).toURL();
            builder.setConnectionFactory(new DefaultConnectionFactory() {
                @Override
                public HttpURLConnection openConnection(final URL originalUrl) throws IOException {
                    // test if the URL is built correctly, ie following the `host` setting
                    if (originalUrl.getHost().equals(endpointUrl.getHost()) && originalUrl.getPort() == endpointUrl.getPort()
                            && originalUrl.getProtocol().equals(endpointUrl.getProtocol())) {
                        return super.openConnection(originalUrl);
                    }
                    // override connection URLs because some don't follow the config. See
                    // https://github.com/GoogleCloudPlatform/google-cloud-java/issues/3254 and
                    // https://github.com/GoogleCloudPlatform/google-cloud-java/issues/3255
                    URI originalUri;
                    try {
                        originalUri = originalUrl.toURI();
                    } catch (final URISyntaxException e) {
                        throw new RuntimeException(e);
                    }
                    String overridePath = "/";
                    if (originalUri.getRawPath() != null) {
                        overridePath = originalUri.getRawPath();
                    }
                    if (originalUri.getRawQuery() != null) {
                        overridePath += "?" + originalUri.getRawQuery();
                    }
                    return super.openConnection(
                            new URL(endpointUrl.getProtocol(), endpointUrl.getHost(), endpointUrl.getPort(), overridePath));
                }
            });
        }
        return builder.build();
    }
```

albertzaharovits commented:
@tlrx I have profiled the _snapshot API with w/ & w/o the multipart update trick. I have attached the flight recordings (profile data, in unpretentious lingo).
resumable & mutipart:
https://drive.google.com/file/d/1KZc4Ua2r8oAttrSMvso7oTD8EirPV1_1/view?usp=sharing
https://drive.google.com/file/d/1wDuTTnqtDYdpZ1TJr3j1sWTaT9BOMnDQ/view?usp=sharing
resumable only:
https://drive.google.com/file/d/1Va_iTVwFrP99Iy96pMGITJrfwxGdxP2a/view?usp=sharing
https://drive.google.com/file/d/1krSQPOiJnpLc4b3bTbXXYHQ3Q3GxtA8n/view?usp=sharing
The VM is looping `PUT _snapshot` and `DELETE _snapshot` for the geonames index (from the rally track) with true GCS credentials. I cannot eye any difference between thee two methods. It might be that the bottleneck is outside the VM.
I have collected the timings too, but they seem un-illustrative, they have variance and it does not appear to be any statistical difference between the snapshot times. I can share them if the profile data is not enough.

Curious if you spot something on the profiling data!

albertzaharovits commented on plugins/repository-gcs/qa/google-cloud-storage/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageTestServer.java:
```diff
@@ -224,33 +199,157 @@ public Response handle(final String method,
+                if (bucket == null) {
+                    return newError(RestStatus.NOT_FOUND, "bucket not found");
+                }
+                if (bucket.objects.put(objectName, EMPTY_BYTE) == null) {
```
nice catch

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettings.java:
```diff
@@ -83,26 +107,37 @@
     }
 
+    public String getProjectId() {
+        return Strings.hasLength(projectId) ? projectId : (credential != null ? credential.getProjectId() : null);
```
nice catch

albertzaharovits commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -241,163 +273,53 @@ void deleteBlobsByPrefix(String prefix) throws IOException {
-
-        if (blobNames.size() == 1) {
-            deleteBlob(blobNames.iterator().next());
+        if (blobNames.size() < 2) {
```
oke

tlrx commented:
I looked at the profiling results from @albertzaharovits and I didn't see any strong differences except that the tests using only the resumable upload method show  more socket I/O, which is expected because this method is supposed to execute more requests. I didn't see any differences in total execution time when snapshotting a large index using resumable and resumable+multipart using the "new" client.

I also compared the snapshot execution time with the "old" client and this "new" client and didn't see any performance change. 

Of course, we should have real benchmarks for this.
